### PR TITLE
qemu_vm: Avoid resume vm in "inmigrate" status

### PR DESCRIPTION
**Pull request by yanan-fu**
After migration, the dst guest might in "inmigration" state for certain
period, especially with "exec:gzip". But resume vm in "inmigrate" can't
work, so wait up to 120s to wait for the right status.

id: 1791785
Signed-off-by: Yanan Fu <yfu@redhat.com>

### CODE REVIEW

**Comment by yanan-fu** [comment]
@ldoktor could you help review ? thanks!

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -4512,7 +4528,10 @@ def resume(self, timeout=None):
         """
         self.monitor.cmd("cont")
         if timeout:
-            self.wait_for_status('running', timeout, 0.1)
+            if not self.wait_for_status('running', timeout, 0.1):
+                raise virt_vm.VMStatusError("Failed to enter running status, "
+                                            "the actual status is %s" %
+                                            self.monitor.get_status())
```
Comment (ldoktor) [inline virttest/qemu_vm.py]: Nice, this is actually a bugfix and not related to the current patch. Would you please extract it as a separate commit?

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -4300,11 +4312,15 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
             # If we're doing remote migration and it's completed successfully,
             # self points to a dead VM object
             if not not_wait_for_migration:
-                if self.is_alive() and self.is_paused():
+                if self.is_alive():
                     # For short period of time the status can be "inmigrate"
                     # for example when using external program
                     # (qemu commit fe823b6f87b2ebedd692ca480ceb9693439d816e)
-                    self.resume(60)
+                    # resume vm during "inmigrate" status can't work
```
Comment (ldoktor) [inline virttest/qemu_vm.py]: The explanation from the commit message is way better. How about something like: `After migration we need to wait for the VM to initialize before sending the "cont". Wait until the status is not "inmigrate".`

(I know that in 5 years I'll be looking at this thinking why are we waiting and why is there inmigrate status...)

Comment (ldoktor) [reply virttest/qemu_vm.py]: Oh stupid me... the comment about inmigrate is here...

Comment (ldoktor) [reply virttest/qemu_vm.py]: well is this patch actually required? We are already waiting for the "cont" to pass for up to 60s with 0.1s step.

File: virttest/qemu_vm.py [hunk 3]
```diff
@@ -4300,11 +4312,15 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
...
+                    if utils_misc.wait_for(lambda: not self.is_inmigrate(),
+                                           timeout=60, step=1):
+                        if self.is_paused():
+                            self.resume()
```
Comment (ldoktor) [inline virttest/qemu_vm.py]: in case it's still in `inmigrate` status you want to ignore it and silently proceed? I think an exception would be more appropriate so we can identify why it stayed in `inmigrate` status. In case it's common we should at least log a message...

File: virttest/qemu_vm.py (hunk 1 above)
Comment (yanan-fu) [reply virttest/qemu_vm.py]: Yes, that is, make it as an separate commit is better.

File: virttest/qemu_vm.py (hunk 2 above)
Comment (yanan-fu) [reply virttest/qemu_vm.py]: @ldoktor  Actually, if we send "cont" in "inmigrate" status, it does not work. Only send "cont" in "paused" status,  vm can be resumed.
This patch will wait the vm status change from "inmigrate" to "paused", then send "cont".
For some other test scenario, vm's status will change from "inmigrate" to "running", in this case, do nothing.

**Comment by yanan-fu** [comment]
What do you mean `it should already work well with a few ignored conts ` ? 
From my checking, only one "cont", and wait for 60s.

File: virttest/qemu_vm.py (hunk 2 above)
Comment (yanan-fu) [reply virttest/qemu_vm.py]: > well is this patch actually required? We are already waiting for the "cont" to pass for up to 60s with 0.1s step.
A slight problem, it is wait for 0.1s before the first checking, let me update it with `step=0.1`

**Comment by yanan-fu** [comment]
@ldoktor  updated, call for your review again, thanks!

**Comment by yanan-fu** [comment]
@xutian could you help check? thanks!

File: virttest/qemu_vm.py (hunk 3 above)
Comment (xutian) [reply virttest/qemu_vm.py]: it's reasonable to raise exception if VM status is not "inmigrate" after timeout.

**Comment by yanan-fu** [comment]
@luckyh @ldoktor  could you help review ? thanks!

File: virttest/qemu_vm.py [hunk 4]
```diff
@@ -4300,11 +4300,19 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
...
-                    self.resume(60)
+                    # resume vm during "inmigrate" status can't work, wait until
+                    # exit "inmigrate" status and resume vm if needed.
+                    if utils_misc.wait_for(lambda: not self.monitor.verify_status('inmigrate'),
+                                           timeout=120, step=0.1):
+                        if self.is_paused():
+                            self.resume()
+                    else:
+                        raise virt_vm.VMStatusError("vm can't exit 'inmigrate' status"
+                                                    "after 120s")
```
Comment (ldoktor) [inline virttest/qemu_vm.py]: It'd be probably useful to log the status here, but it's not that important as we can get it from the monitor logs...

Comment (yanan-fu) [reply virttest/qemu_vm.py]: @ldoktor From the error info in debug.log: `can't exit 'inmigrate' status`, it means the status is "inmigrate" after 120s.
So i think we do not need to log `self.monitor.get_status()` again. Thanks!

**Comment by xutian** [comment]
@yanan-fu   typo in commit message   sed  's/retuen/return'

File: virttest/qemu_vm.py [hunk 5]
```diff
@@ -4300,11 +4300,19 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
...
                     # For short period of time the status can be "inmigrate"
                     # for example when using external program
                     # (qemu commit fe823b6f87b2ebedd692ca480ceb9693439d816e)
-                    self.resume(60)
+                    # resume vm during "inmigrate" status can't work, wait until
+                    # exit "inmigrate" status and resume vm if needed.
+                    if utils_misc.wait_for(lambda: not self.monitor.verify_status('inmigrate'),
+                                           timeout=120, step=0.1):
+                        if self.is_paused():
+                            self.resume()
```
Comment (xutian) [inline virttest/qemu_vm.py]: do we need timeout for this resume?

Comment (yanan-fu) [reply virttest/qemu_vm.py]: @xutian Thanks, i think it is no need. resume timeout was involved by commit 61dcaa106b5e6a21c3fa7931a35f20374112ff8c, it aim to send "cont" one time then wait for the vm status change from inmigrate to running.  But in fact, send "cont" while vm status is inmigrate, and with "-S",  it can not resume the vm. 
With this patch, it will send 'cont' after vm exit 'inmigrate' status. no need to add timeout.

**Comment by yanan-fu** [comment]
> @yanan-fu typo in commit message sed 's/retuen/return'
 Thanks, updated ~

**Comment by yanan-fu** [comment]
can reproduce this issue with test case "migrate.default.exec.gzip_exec.default" and set "offline=yes"
Test pass now:
 (1/1) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.migrate.exec.gzip_exec: PASS (200.97 s)

**Comment by xutian** [comment]
merged, thanks @yanan-fu and @ldoktor
